## Table of Contents

- [Introduction](#introduction)
  - [10x Genomics V(D)J Platform](#10x-genomics-vdj-platform)
  - [Running Cell Ranger vdj](#running-cell-ranger-vdj)
    - [Data](#data)
    - [Cell Ranger](#cell-ranger)
    - [AIRR rearrangements TSV file](#airr-rearrangements-tsv-file)
- [Adaptive Immune Receptor Repertoire analysis](#adaptive-immune-receptor-repertoire-analysis)
  - [Setup](#setup)
  - [Test results](#test-results)
  - [Input samplesheet](#input-samplesheet)
    - [Bulk](#bulk)
    - [Single cell](#single-cell)
    - [Assembled input samplesheet](#assembled-input-samplesheet)
    - [Troubleshooting](#troubleshooting)

# Introduction

* Immune repertoire analysis is the study of the **diversity** and **composition** of antigen receptors expressed by B Cells Receptors (BCRs) and T Cells Receptors (TCRs).
* By sequencing the Variable (V), Diversity (D), and Joining (J) gene segments of these receptors, we can assess the clonal diversity, expansion, and specificity of immune cells in different conditions.

## 10x Genomics V(D)J Platform

* The **10x Genomics Chromium Single Cell Immune Profiling** platform allows high-throughput single-cell sequencing of V(D)J regions for both BCRs and TCRs.
* Antigen receptor sequences are typically linked to single-cell gene expression data.
* The **Chromium Controller** encapsulates single cells into droplets, each containing a unique barcoded bead.
    * cDNAs are labelled during reverse transcription.
* Full-length V(D)J transcripts  are amplified from single cells.
* Construct libraries separately for TCR or BCR sequences.
* Cell Ranger is used to aligns reads, reconstructs full-length V(D)J sequences, and clusters clonotypes.

## Running Cell Ranger vdj

[A tutorial on running the cellranger vdj pipeline](https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/tutorial/tutorial-vdj).

Cell Ranger's vdj pipeline

The cellranger vdj pipeline takes FASTQ files from `cellranger mkfastq`, `BCL Convert`, or `bcl2fastq` for V(D)J libraries and performs sequence assembly and paired clonotype calling. It uses the Chromium cellular barcodes and UMIs to assemble V(D)J transcripts per cell. Clonotypes and CDR3 sequences are output as a `.vloupe` file which can be loaded into the Loupe V(D)J Browser.

### Data

[Human B cells from a Healthy Donor, 1k cells](https://www.10xgenomics.com/datasets/human-b-cells-from-a-healthy-donor-1-k-cells-2-standard-6-0-0)

* Universal 5' Gene Expression dataset analyzed using Cell Ranger 6.0.0
* Human CD19+ B cells of a healthy male donor aged 27 were obtained by 10x Genomics from AllCells.
* Gene Expression and BCR Enriched libraries were generated from 1,650 cells (89% viable, expected cell recovery 915) as described in the Chromium Next GEM Single Cell 5' Reagent Kits v2 (Dual Index) User Guide (CG000331 Rev A) and sequenced on an Illumina NovaSeq 6000.
* These libraries were analyzed using the new Cell Ranger multi pipeline, which combines Gene Expression, Feature Barcode, TCR enriched, and BCR enriched libraries from the same sample. Cell Ranger multi enables consistent cell calling between gene expression and VDJ libraries.
* Sequencing Configuration - Paired-end (26X90), dual-indexed sequencing. Read 1: 26 cycles, i7 index: 10 cycles, i5 index: 10 cycles, Read 2: 90 cycles
* This dataset is licensed under the Creative Commons Attribution 4.0 International (CC BY 4.0) license. 10x citation guidelines available here.

FASTQ indexes.

```console
zcat sc5p_v2_hs_B_1k_b_S1_L001_I1_001.fastq.gz | head -4
```
```
@A00984:155:HNW5YDMXX:1:1101:12970:1000 1:N:0:ACCAGACAAC+AGGAACTAGG
ACCAGACAAC
+
FFFFFFFF:F
```
```console
zcat sc5p_v2_hs_B_1k_b_S1_L001_I2_001.fastq.gz | head -4
```
```
@A00984:155:HNW5YDMXX:1:1101:12970:1000 2:N:0:ACCAGACAAC+AGGAACTAGG
AGGAACTAGG
+
,FFF:FFFFF
```

FASTQ sequences.

```console
zcat sc5p_v2_hs_B_1k_b_S1_L001_R1_001.fastq.gz | head -4
```
```
@A00984:155:HNW5YDMXX:1:1101:12970:1000 1:N:0:ACCAGACAAC+AGGAACTAGG
GTCATTTGTCTTTCATAGACGTCGGT
+
FFFFFFFFFFFFFFFF,FFFFFFFFF
```
```console
zcat sc5p_v2_hs_B_1k_b_S1_L001_R2_001.fastq.gz | head -4
```
```
@A00984:155:HNW5YDMXX:1:1101:12970:1000 2:N:0:ACCAGACAAC+AGGAACTAGG
CTACACCTCTCATGTCTCTTCACCGTGCCAGACTAGAGTCAAGCTCAACAGGGTCTTCTTTCCCCGCTGATTCCGCCAAGCCCGTTCCCT
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:F:FFFFFF:FFFFFFFFFFFFFFFFFFFFFFFF
```

### Cell Ranger

Run Cell Ranger vdj.

```console
cellranger vdj \
    --id=HumanB_Cell \
    --reference=../refdata-cellranger-vdj-GRCh38-alts-ensembl-7.1.0 \
    --fastqs=../dataset-vdj-practice/sc5p_v2_hs_B_1k_multi_5gex_b_fastqs/sc5p_v2_hs_B_1k_b_fastqs \
    --sample=sc5p_v2_hs_B_1k_b \
    --localcores=4 \
    --localmem=60
```

Output will in a directory with the same name as the argument used for `--id`, e.g., `HumanB_Cell`.

```console
tree --charset ascii -L 1 HumanB_Cell
```
```
HumanB_Cell
|-- _cmdline
|-- extras
|-- _filelist
|-- _finalstate
|-- HumanB_Cell.mri.tgz
|-- _invocation
|-- _jobmode
|-- _log
|-- _mrosource
|-- outs
|-- _perf
|-- _perf._truncated_
|-- SC_VDJ_ASSEMBLER_CS
|-- _sitecheck
|-- _tags
|-- _timestamp
|-- _uuid
|-- _vdrkill
`-- _versions
```

Outputs in `outs`.

```console
tree --charset ascii -L 1 HumanB_Cell/outs
```
```
HumanB_Cell/outs
|-- airr_rearrangement.tsv
|-- all_contig_annotations.bed
|-- all_contig_annotations.csv
|-- all_contig_annotations.json
|-- all_contig.bam
|-- all_contig.bam.bai
|-- all_contig.fasta
|-- all_contig.fasta.fai
|-- all_contig.fastq
|-- cell_barcodes.json
|-- clonotypes.csv
|-- concat_ref.bam
|-- concat_ref.bam.bai
|-- concat_ref.fasta
|-- concat_ref.fasta.fai
|-- consensus_annotations.csv
|-- consensus.bam
|-- consensus.bam.bai
|-- consensus.fasta
|-- consensus.fasta.fai
|-- donor_regions.fa
|-- filtered_contig_annotations.csv
|-- filtered_contig.fasta
|-- filtered_contig.fastq
|-- metrics_summary.csv
|-- vdj_contig_info.pb
|-- vdj_reference
|-- vloupe.vloupe
`-- web_summary.html
```

The files above (in the `outs` directory) have been uploaded to my web server; use the path `https://davetang.org/file/HumanB_Cell/outs/` and the file name above to access/download them. For example to see the web summary head to <https://davetang.org/file/HumanB_Cell/outs/web_summary.html>.

### AIRR rearrangements TSV file

The `airr_rearrangement.tsv` [file](https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/output/annotation#airr) provides the annotated contigs and consensus sequences of V(D)J rearrangements in the AIRR format. The AIRR rearrangement file includes all mandatory AIRR fields and several optional variables to enhance reproducibility and guide analyses.

| Column             | Description                                                                             |
| -                  | -                                                                                       |
| cell_id            | Cell barcode defining the cell for the query sequence.                                  |
| clone_id           | Clonotype ID/clonotype assignment.                                                      |
| sequence_id        | The name of the contig associated with the rearrangement.                               |
| sequence           | The nucleotide sequence of the rearrangement.                                           |
| sequence_aa        | The amino acid sequence of the rearrangement.                                           |
| productive         | Whether or not the rearrangement is productive.                                         |
| rev_comp           | Set to false by default (10x Genomics V(D)J sequences are not reverse complemented).    |
| v_call             | The name of the aligned V gene for the rearrangement.                                   |
| v_cigar            | The CIGAR string of the V gene alignment.                                               |
| d_call             | The name of the aligned D gene for the rearrangement.                                   |
| d_cigar            | The CIGAR string of the D gene alignment.                                               |
| j_call             | The name of the aligned J gene for the rearrangement.                                   |
| j_cigar            | The CIGAR string of the J gene alignment.                                               |
| c_call             | The name of the aligned C gene for the rearrangement.                                   |
| c_cigar            | The CIGAR string of the C gene alignment.                                               |
| sequence_alignment | The aligned sequence of the VDJ rearrangement.                                          |
| germline_alignment | The assembled, aligned, full-length inferred germline sequence of the aligned sequence. |
| junction           | The nucleotide sequence of the rearrangement's junction (CDR3).                         |
| junction_aa        | The amino acid sequence of the rearrangement's junction (CDR3).                         |
| junction_length    | The length of the rearrangement's junction nucleotide sequence.                         |
| junction_aa_length | The length of the rearrangement's junction amino acid sequence.                         |
| v_sequence_start   | 1-based index on the contig of the V region start position.                             |
| v_sequence_end     | 1-based index on the contig of the V region end position.                               |
| d_sequence_start   | 1-based index on the contig of the D region start position.                             |
| d_sequence_end     | 1-based index on the contig of the D region end position.                               |
| j_sequence_start   | 1-based index on the contig of the J region start position.                             |
| j_sequence_end     | 1-based index on the contig of the J region end position.                               |
| c_sequence_start   | 1-based index on the contig of the C region start position.                             |
| c_sequence_end     | 1-based index on the contig of the C region end position.                               |
| consensus_count    | The number of reads associated with this rearrangement.                                 |
| duplicate_count    | The number of unique molecular identifiers associated with this rearrangement.          |
| is_cell            | Is this rearrangement cell-associated?                                                  |

An example `airr_rearrangement.tsv` file.

```console
wget -q -O - https://davetang.org/file/HumanB_Cell/outs/airr_rearrangement.tsv | head
```
```
cell_id	clone_id	sequence_id	sequence	sequence_aa	productive	rev_comp	v_call	v_cigar	d_call	d_cigar	j_call	j_cigar	c_call	c_cigar	sequence_alignment	germline_alignment	junction	junction_aa	junction_length	junction_aa_length	v_sequence_start	v_sequence_end	d_sequence_start	d_sequence_end	j_sequence_start	j_sequence_end	c_sequence_start	c_sequence_end	consensus_count	duplicate_count	is_cell
AAACCTGAGGGCTCTC-1	clonotype607	AAACCTGAGGGCTCTC-1_contig_1	ACTTTCTGAGAGTCCTGGACCTCCTGTGCAAGAACATGAAACATCTGTGGTTCTTCCTCCTCCTGGTGGCAGCTCCCAGATGGGTCCTGTCCCAGGTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCACAGACCCTGTCCCTCACCTGCACTGTCTCTGGTGGCTCCATCAGCAGTGGTAGTTACTACTGGAGCTGGATCCGGCAGCCCGCCGGGAAGGGACTGGAGTGGATTGGGCGTATCTATACCAGTGGGAGCACCAACTACAACCCCTCCCTCAAGAGTCGAGTCACCATATCAGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTCTGTGACCGCCGCAGACACGGCCGTGTATTACTGTGCGAGAGGGGATAGCAGTGGCTGGCGAGGGGGCAACTGGTTCGACCCCTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	MKHLWFFLLLVAAPRWVLSQVQLQESGPGLVKPSQTLSLTCTVSGGSISSGSYYWSWIRQPAGKGLEWIGRIYTSGSTNYNPSLKSRVTISVDTSKNQFSLKLSSVTAADTAVYYCARGDSSGWRGGNWFDPWGQGTLVTVSSGSASAPTLFPLVSCENSPSDTSSV	T	F	IGHV4-61	35S356M145S			IGHJ5	414S51M71S	IGHM	465S71M	ACTTTCTGAGAGTCCTGGACCTCCTGTGCAAGAACATGAAACATCTGTGGTTCTTCCTCCTCCTGGTGGCAGCTCCCAGATGGGTCCTGTCCCAGGTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCACAGACCCTGTCCCTCACCTGCACTGTCTCTGGTGGCTCCATCAGCAGTGGTAGTTACTACTGGAGCTGGATCCGGCAGCCCGCCGGGAAGGGACTGGAGTGGATTGGGCGTATCTATACCAGTGGGAGCACCAACTACAACCCCTCCCTCAAGAGTCGAGTCACCATATCAGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTCTGTGACCGCCGCAGACACGGCCGTGTATTACTGTGCGAGAGGGGATAGCAGTGGCTGGCGAGGGGGCAACTGGTTCGACCCCTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	ACTTAAGCACCCACAGGAAACCACCACACATTTCCTTAAATTCAGGTTCCAGCTCACATGGGAAATACTTTCTGAGAGTCCTGGACCTCCTGTGCAAGAACATGAAACACCTGTGGTTCTTCCTCCTCCTGGTGGCAGCTCCCAGATGGGTCCTGTCCCAGGTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGGAGACCCTGTCCCTCACCTGCACTGTCTCTGGTGGCTCCGTCAGCAGTGGTAGTTACTACTGGAGCTGGATCCGGCAGCCCCCAGGGAAGGGACTGGAGTGGATTGGGTATATCTATTACAGTGGGAGCACCAACTACAACCCCTCCCTCAAGAGTCGAGTCACCATATCAGTAGACACGTCCAAGAACCAGTTCTCCCTGAAGCTGAGCTCTGTGACCGCTGCGGACACGGCCGTGTATTACTGTGCGAGAGAACAACTGGTTCGACCCCTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTGGCCGTTGGCTGCCTCGCACAGGACTTCCTTCCCGACTCCATCACTTTCTCCTGGAAATACAAGAACAACTCTGACATCAGCAGCACCCGGGGCTTCCCATCAGTCCTGAGAGGGGGCAAGTACGCAGCCACCTCACAGGTGCTGCTGCCTTCCAAGGACGTCATGCAGGGCACAGACGAACACGTGGTGTGCAAAGTCCAGCACCCCAACGGCAACAAAGAAAAGAACGTGCCTCTTCCAGTGATTGCTGAGCTGCCTCCCAAAGTGAGCGTCTTCGTCCCACCCCGCGACGGCTTCTTCGGCAACCCCCGCAAGTCCAAGCTCATCTGCCAGGCCACGGGTTTCAGTCCCCGGCAGATTCAGGTGTCCTGGCTGCGCGAGGGGAAGCAGGTGGGGTCTGGCGTCACCACGGACCAGGTGCAGGCTGAGGCCAAAGAGTCTGGGCCCACGACCTACAAGGTGACCAGCACACTGACCATCAAAGAGAGCGACTGGCTCGGCCAGAGCATGTTCACCTGCCGCGTGGATCACAGGGGCCTGACCTTCCAGCAGAATGCGTCCTCCATGTGTGTCCCCGATCAAGACACAGCCATCCGGGTCTTCGCCATCCCCCCATCCTTTGCCAGCATCTTCCTCACCAAGTCCACCAAGTTGACCTGCCTGGTCACAGACCTGACCACCTATGACAGCGTGACCATCTCCTGGACCCGCCAGAATGGCGAAGCTGTGAAAACCCACACCAACATCTCCGAGAGCCACCCCAATGCCACTTTCAGCGCCGTGGGTGAGGCCAGCATCTGCGAGGATGACTGGAATTCCGGGGAGAGGTTCACGTGCACCGTGACCCACACAGACCTGCCCTCGCCACTGAAGCAGACCATCTCCCGGCCCAAGGGGGTGGCCCTGCACAGGCCCGATGTCTACTTGCTGCCACCAGCCCGGGAGCAGCTGAACCTGCGGGAGTCGGCCACCATCACGTGCCTGGTGACGGGCTTCTCTCCCGCGGACGTCTTCGTGCAGTGGATGCAGAGGGGGCAGCCCTTGTCCCCGGAGAAGTATGTGACCAGCGCCCCAATGCCTGAGCCCCAGGCCCCAGGCCGGTACTTCGCCCACAGCATCCTGACCGTGTCCGAAGAGGAATGGAACACGGGGGAGACCTACACCTGCGTGGTGGCCCATGAGGCCCTGCCCAACAGGGTCACCGAGAGGACCGTGGACAAGTCCACCGAGGGGGAGGTGAGCGCCGACGAGGAGGGCTTTGAGAACCTGTGGGCCACCGCCTCCACCTTCATCGTCCTCTTCCTCCTGAGCCTCTTCTACAGTACCACCGTCACCTTGTTCAAGGTGAAA	TGTGCGAGAGGGGATAGCAGTGGCTGGCGAGGGGGCAACTGGTTCGACCCCTGG	CARGDSSGWRGGNWFDPW	54	18	36	391			415	465	466	536	889	14	T
AAACCTGAGGGCTCTC-1	clonotype607	AAACCTGAGGGCTCTC-1_contig_2	GCTCTGCTTCAGCTGTGGGCACAAGAGGCAGCACTCAGGACAATCTCCAGCATGGCCTGGTCTCCTCTCCTCCTCACTCTCCTCGCTCACTGCACAGGGTCCTGGGCCCAGTCTGTGCTGACGCAGCCGCCCTCAGTGTCTGGGGCCCCAGGGCAGAGGGTCACCATCTCCTGCACTGGGAGCAGCTCCAACATCGGGGCAGGTTATGATGTACACTGGTACCAGCAGCTTCCAGGAACAGCCCCCAAACTCCTCATCTATGGTAACAGCAATCGGCCCTCAGGGGTCCCTGACCGATTCTCTGGCTCCAAGTCTGGCACCTCAGCCTCCCTGGCCATCACTGGGCTCCAGGCTGAGGATGAGGCTGATTATTACTGCCAGTCCTATGACAGCAGCCTGAGTGACGTGTTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCACCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTA	MAWSPLLLTLLAHCTGSWAQSVLTQPPSVSGAPGQRVTISCTGSSSNIGAGYDVHWYQQLPGTAPKLLIYGNSNRPSGVPDRFSGSKSGTSASLAITGLQAEDEADYYCQSYDSSLSDVFGGGTKLTVLGQPKAAPSVTLFPPSSEELQANKATLVCLISDFYPGAVTVAWKADSSPVKAGVETTTPSKQSNNKYAASS	T	F	IGLV1-40	51S356M243S			IGLJ2	401S38M211S	IGLC3	439S211M	GCTCTGCTTCAGCTGTGGGCACAAGAGGCAGCACTCAGGACAATCTCCAGCATGGCCTGGTCTCCTCTCCTCCTCACTCTCCTCGCTCACTGCACAGGGTCCTGGGCCCAGTCTGTGCTGACGCAGCCGCCCTCAGTGTCTGGGGCCCCAGGGCAGAGGGTCACCATCTCCTGCACTGGGAGCAGCTCCAACATCGGGGCAGGTTATGATGTACACTGGTACCAGCAGCTTCCAGGAACAGCCCCCAAACTCCTCATCTATGGTAACAGCAATCGGCCCTCAGGGGTCCCTGACCGATTCTCTGGCTCCAAGTCTGGCACCTCAGCCTCCCTGGCCATCACTGGGCTCCAGGCTGAGGATGAGGCTGATTATTACTGCCAGTCCTATGACAGCAGCCTGAGTGACGTGTTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCACCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTA	GCTCTGCTTCAGCTGTGGGCACAAGAGGCAGCACTCAGGACAATCTCCAGCATGGCCTGGTCTCCTCTCCTCCTCACTCTCCTCGCTCACTGCACAGGGTCCTGGGCCCAGTCTGTGCTGACGCAGCCGCCCTCAGTGTCTGGGGCCCCAGGGCAGAGGGTCACCATCTCCTGCACTGGGAGCAGCTCCAACATCGGGGCAGGTTATGATGTACACTGGTACCAGCAGCTTCCAGGAACAGCCCCCAAACTCCTCATCTATGGTAACAGCAATCGGCCCTCAGGGGTCCCTGACCGATTCTCTGGCTCCAAGTCTGGCACCTCAGCCTCCCTGGCCATCACTGGGCTCCAGGCTGAGGATGAGGCTGATTATTACTGCCAGTCCTATGACAGCAGCCTGAGTGGTTCTGTGGTATTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCACCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTACCTGAGCCTGACGCCTGAGCAGTGGAAGTCCCACAAAAGCTACAGCTGCCAGGTCACGCATGAAGGGAGCACCGTGGAGAAGACAGTGGCCCCTACAGAATGTTCA	TGCCAGTCCTATGACAGCAGCCTGAGTGACGTGTTC	CQSYDSSLSDVF	36	12	52	407			402	439	440	650	5214	60	T
AAACCTGGTAAGGATT-1	clonotype660	AAACCTGGTAAGGATT-1_contig_1	GAGCTACAACAGGCAGGCAGGGGCAGCAAGATGGTGTTGCAGACCCAGGTCTTCATTTCTCTGTTGCTCTGGATCTCTGGTGCCTCCGGGGACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGGGCGAGAGGGCCACCGTCAACTGCAAGTCCAGCCAGAGTGTTTTATACAGCTCCAACAAGAAGAACTACTTAGCTTGGTACCAGCAGAAACCAGGACAGCCTCCTAAATTGCTCCTTTACTGGGCATCTACCCGGGAATCCGGGGTCCCTGACCGGTTCAGTGGCAGCGGGTCTGGGACAGATTTCACTCTCACCATCAGCAGCCTGCAGGCTGAAGATGTGGCTGTTTATTACTGTCAGCAATACTATGATACTCCCCGTACTTTTGGCCAGGGGACCAGGCTGGAGATCAAACGAACTGTGGCTGCACCATCTGTCTTCATCTTCCCGCCATCTGATGAGCAGTTGAAATCTGGAACTGCCTCTGTTGTGTGCCTGCTGAATAACTTCTATCCCAGAGAGGCCAAAGTACAGTGGAAGGTGGATAACGC	MVLQTQVFISLLLWISGASGDIVMTQSPDSLAVSLGERATVNCKSSQSVLYSSNKKNYLAWYQQKPGQPPKLLLYWASTRESGVPDRFSGSGSGTDFTLTISSLQAEDVAVYYCQQYYDTPRTFGQGTRLEIKRTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDN	T	F	IGKV4-1	30S363M173S			IGKJ2	391S39M136S	IGKC	430S136M	GAGCTACAACAGGCAGGCAGGGGCAGCAAGATGGTGTTGCAGACCCAGGTCTTCATTTCTCTGTTGCTCTGGATCTCTGGTGCCTCCGGGGACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGGGCGAGAGGGCCACCGTCAACTGCAAGTCCAGCCAGAGTGTTTTATACAGCTCCAACAAGAAGAACTACTTAGCTTGGTACCAGCAGAAACCAGGACAGCCTCCTAAATTGCTCCTTTACTGGGCATCTACCCGGGAATCCGGGGTCCCTGACCGGTTCAGTGGCAGCGGGTCTGGGACAGATTTCACTCTCACCATCAGCAGCCTGCAGGCTGAAGATGTGGCTGTTTATTACTGTCAGCAATACTATGATACTCCCCGTACTTTTGGCCAGGGGACCAGGCTGGAGATCAAACGAACTGTGGCTGCACCATCTGTCTTCATCTTCCCGCCATCTGATGAGCAGTTGAAATCTGGAACTGCCTCTGTTGTGTGCCTGCTGAATAACTTCTATCCCAGAGAGGCCAAAGTACAGTGGAAGGTGGATAACGC	TTTGGCTCTTGATTTACATTGGGTACTTTCACAACCCACTGCTCATGAAATTTGCTTTTGTACTCACTGGTTGTTTTTGCATAGGCCCCTCCAGGCCACGACCAGCTGTTTGGATTTTATAAACGGGCCGTTTGCATTGTGAACTGAGCTACAACAGGCAGGCAGGGGCAGCAAGATGGTGTTGCAGACCCAGGTCTTCATTTCTCTGTTGCTCTGGATCTCTGGTGCCTACGGGGACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGGGCGAGAGGGCCACCATCAACTGCAAGTCCAGCCAGAGTGTTTTATACAGCTCCAACAATAAGAACTACTTAGCTTGGTACCAGCAGAAACCAGGACAGCCTCCTAAGCTGCTCATTTACTGGGCATCTACCCGGGAATCCGGGGTCCCTGACCGATTCAGTGGCAGCGGGTCTGGGACAGATTTCACTCTCACCATCAGCAGCCTGCAGGCTGAAGATGTGGCAGTTTATTACTGTCAGCAATATTATAGTACTCCTTGTGCAGTTTTGGCCAGGGGACCAAGCTGGAGATCAAACGAACTGTGGCTGCACCATCTGTCTTCATCTTCCCGCCATCTGATGAGCAGTTGAAATCTGGAACTGCCTCTGTTGTGTGCCTGCTGAATAACTTCTATCCCAGAGAGGCCAAAGTACAGTGGAAGGTGGATAACGCCCTCCAATCGGGTAACTCCCAGGAGAGTGTCACAGAGCAGGACAGCAAGGACAGCACCTACAGCCTCAGCAGCACCCTGACGCTGAGCAAAGCAGACTACGAGAAACACAAAGTCTACGCCTGCGAAGTCACCCATCAGGGCCTGAGCTCGCCCGTCACAAAGAGCTTCAACAGGGGAGAGTGT	TGTCAGCAATACTATGATACTCCCCGTACTTTT	CQQYYDTPRTF	33	11	31	393			392	430	431	566	2099	28	T
AAACCTGGTAAGGATT-1	clonotype660	AAACCTGGTAAGGATT-1_contig_2	GGGAGCATCACCCAGCAACCACATCTGTCCTCTAGAGAATCCCCTGAGAGCTCCGTTCCTCACCATGGACTGGACCTGGAGGATCCTCTTCTTGGTGGCAGCAGCCACAGGAGGCCACTCACAGGTGCAGCTGGTGCAGTCTGGGGCTGAGGTGAAGAAGCCTGGGGCCTCAGTGAAGGTCTCCTGCAAGGCTTCTGGATACACCCTCACCGGCTACTATATACACTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAATACTAATAGCGGTGGCACAAACTATGCACAGAAGTTTCAGGGCAGGGTCACCATGACCAGGGACACGTCCATCAGCACAGCCTACATGGAGGTGAGCGGCCTGAGATCTGACGACACGGCCGTGTATTATTGTGCGATGGGATATTGTATTAATAATAACTGTTACGAGGGGTGGTTCGACCCCTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	MDWTWRILFLVAAATGGHSQVQLVQSGAEVKKPGASVKVSCKASGYTLTGYYIHWVRQAPGQGLEWMGWINTNSGGTNYAQKFQGRVTMTRDTSISTAYMEVSGLRSDDTAVYYCAMGYCINNNCYEGWFDPWGQGTLVTVSSGSASAPTLFPLVSCENSPSDTSSV	T	F	IGHV1-2	64S353M148S			IGHJ5	443S51M71S	IGHM	494S71M	GGGAGCATCACCCAGCAACCACATCTGTCCTCTAGAGAATCCCCTGAGAGCTCCGTTCCTCACCATGGACTGGACCTGGAGGATCCTCTTCTTGGTGGCAGCAGCCACAGGAGGCCACTCACAGGTGCAGCTGGTGCAGTCTGGGGCTGAGGTGAAGAAGCCTGGGGCCTCAGTGAAGGTCTCCTGCAAGGCTTCTGGATACACCCTCACCGGCTACTATATACACTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAATACTAATAGCGGTGGCACAAACTATGCACAGAAGTTTCAGGGCAGGGTCACCATGACCAGGGACACGTCCATCAGCACAGCCTACATGGAGGTGAGCGGCCTGAGATCTGACGACACGGCCGTGTATTATTGTGCGATGGGATATTGTATTAATAATAACTGTTACGAGGGGTGGTTCGACCCCTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	GAGAGCATCACCCAGCAACCACATCTGTCCTCTAGAGAATCCCCTGAGAGCTCCGTTCCTCACCATGGACTGGACCTGGAGGATCCTCTTCTTGGTGGCAGCAGCCACAGGAGCCCACTCCCAGGTGCAGCTGGTGCAGTCTGGGGCTGAGGTGAAGAAGCCTGGGGCCTCAGTGAAGGTCTCCTGCAAGGCTTCTGGATACACCTTCACCGGCTACTATATGCACTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAACCCTAACAGTGGTGGCACAAACTATGCACAGAAGTTTCAGGGCTGGGTCACCATGACCAGGGACACGTCCATCAGCACAGCCTACATGGAGCTGAGCAGGCTGAGATCTGACGACACGGCCGTGTATTACTGTGCGAGAGAACAACTGGTTCGACCCCTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTGGCCGTTGGCTGCCTCGCACAGGACTTCCTTCCCGACTCCATCACTTTCTCCTGGAAATACAAGAACAACTCTGACATCAGCAGCACCCGGGGCTTCCCATCAGTCCTGAGAGGGGGCAAGTACGCAGCCACCTCACAGGTGCTGCTGCCTTCCAAGGACGTCATGCAGGGCACAGACGAACACGTGGTGTGCAAAGTCCAGCACCCCAACGGCAACAAAGAAAAGAACGTGCCTCTTCCAGTGATTGCTGAGCTGCCTCCCAAAGTGAGCGTCTTCGTCCCACCCCGCGACGGCTTCTTCGGCAACCCCCGCAAGTCCAAGCTCATCTGCCAGGCCACGGGTTTCAGTCCCCGGCAGATTCAGGTGTCCTGGCTGCGCGAGGGGAAGCAGGTGGGGTCTGGCGTCACCACGGACCAGGTGCAGGCTGAGGCCAAAGAGTCTGGGCCCACGACCTACAAGGTGACCAGCACACTGACCATCAAAGAGAGCGACTGGCTCGGCCAGAGCATGTTCACCTGCCGCGTGGATCACAGGGGCCTGACCTTCCAGCAGAATGCGTCCTCCATGTGTGTCCCCGATCAAGACACAGCCATCCGGGTCTTCGCCATCCCCCCATCCTTTGCCAGCATCTTCCTCACCAAGTCCACCAAGTTGACCTGCCTGGTCACAGACCTGACCACCTATGACAGCGTGACCATCTCCTGGACCCGCCAGAATGGCGAAGCTGTGAAAACCCACACCAACATCTCCGAGAGCCACCCCAATGCCACTTTCAGCGCCGTGGGTGAGGCCAGCATCTGCGAGGATGACTGGAATTCCGGGGAGAGGTTCACGTGCACCGTGACCCACACAGACCTGCCCTCGCCACTGAAGCAGACCATCTCCCGGCCCAAGGGGGTGGCCCTGCACAGGCCCGATGTCTACTTGCTGCCACCAGCCCGGGAGCAGCTGAACCTGCGGGAGTCGGCCACCATCACGTGCCTGGTGACGGGCTTCTCTCCCGCGGACGTCTTCGTGCAGTGGATGCAGAGGGGGCAGCCCTTGTCCCCGGAGAAGTATGTGACCAGCGCCCCAATGCCTGAGCCCCAGGCCCCAGGCCGGTACTTCGCCCACAGCATCCTGACCGTGTCCGAAGAGGAATGGAACACGGGGGAGACCTACACCTGCGTGGTGGCCCATGAGGCCCTGCCCAACAGGGTCACCGAGAGGACCGTGGACAAGTCCACCGAGGGGGAGGTGAGCGCCGACGAGGAGGGCTTTGAGAACCTGTGGGCCACCGCCTCCACCTTCATCGTCCTCTTCCTCCTGAGCCTCTTCTACAGTACCACCGTCACCTTGTTCAAGGTGAAA	TGTGCGATGGGATATTGTATTAATAATAACTGTTACGAGGGGTGGTTCGACCCCTGG	CAMGYCINNNCYEGWFDPW	57	19	65	417			444	494	495	565	1410	20	T
AAACCTGGTAATAGCA-1	clonotype421	AAACCTGGTAATAGCA-1_contig_1	ACCCAAAAACCACACCCCTCCTTGGGAGAATCCCCTAGATCACAGCTCCTCACCATGGACTGGACCTGGAGCATCCTTTTCTTGGTGGCAGCAGCAACAGGTGCCCACTCCCAGGTTCAGCTGGTGCAGTCTGGAGCTGAGGTGAAGAAGCCTGGGGCCTCAGTGAAGGTCTCCTGCAAGGCTTCTGGTTACACCTTTACCAGCTATGGTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAGCGCTTACAATGGTAACACAAACTATGCACAGAAGCTCCAGGGCAGAGTCACCATGACCACAGACACATCCACGAGCACAGCCTACATGGAGCTGAGGAGCCTGAGATCTGACGACACGGCCGTGTATTACTGTGCGAGAGCCAAACGTTGGGGGTATAGCAGCAGCTGGTGCGACTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	MDWTWSILFLVAAATGAHSQVQLVQSGAEVKKPGASVKVSCKASGYTFTSYGISWVRQAPGQGLEWMGWISAYNGNTNYAQKLQGRVTMTTDTSTSTAYMELRSLRSDDTAVYYCARAKRWGYSSSWCDYWGQGTLVTVSSGSASAPTLFPLVSCENSPSDTSSV	T	F	IGHV1-18	54S353M142S	IGHD6-13	417S21M111S	IGHJ5	427S51M71S	IGHM	478S71M	ACCCAAAAACCACACCCCTCCTTGGGAGAATCCCCTAGATCACAGCTCCTCACCATGGACTGGACCTGGAGCATCCTTTTCTTGGTGGCAGCAGCAACAGGTGCCCACTCCCAGGTTCAGCTGGTGCAGTCTGGAGCTGAGGTGAAGAAGCCTGGGGCCTCAGTGAAGGTCTCCTGCAAGGCTTCTGGTTACACCTTTACCAGCTATGGTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAGCGCTTACAATGGTAACACAAACTATGCACAGAAGCTCCAGGGCAGAGTCACCATGACCACAGACACATCCACGAGCACAGCCTACATGGAGCTGAGGAGCCTGAGATCTGACGACACGGCCGTGTATTACTGTGCGAGAGCCAAACGTTGGGGGTATAGCAGCAGCTGGTGCGACTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	GCATCACCCAAAAACCACACCCCTCCTTGGGAGAATCCCCTAGATCACAGCTCCTCACCATGGACTGGACCTGGAGCATCCTTTTCTTGGTGGCAGCAGCAACAGGTGCCCACTCCCAGGTTCAGCTGGTGCAGTCTGGAGCTGAGGTGAAGAAGCCTGGGGCCTCAGTGAAGGTCTCCTGCAAGGCTTCTGGTTACACCTTTACCAGCTACGGTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAGCGCTTACAATGGTAACACAAACTATGCACAGAAGCTCCAGGGCAGAGTCACCATGACCACAGACACATCCACGAGCACAGCCTACATGGAGCTGAGGAGCCTGAGATCTGACGACACGGCCGTGTATTACTGTGCGAGAGAGGGTATAGCAGCAGCTGGTACACAACTGGTTCGACCCCTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTGGCCGTTGGCTGCCTCGCACAGGACTTCCTTCCCGACTCCATCACTTTCTCCTGGAAATACAAGAACAACTCTGACATCAGCAGCACCCGGGGCTTCCCATCAGTCCTGAGAGGGGGCAAGTACGCAGCCACCTCACAGGTGCTGCTGCCTTCCAAGGACGTCATGCAGGGCACAGACGAACACGTGGTGTGCAAAGTCCAGCACCCCAACGGCAACAAAGAAAAGAACGTGCCTCTTCCAGTGATTGCTGAGCTGCCTCCCAAAGTGAGCGTCTTCGTCCCACCCCGCGACGGCTTCTTCGGCAACCCCCGCAAGTCCAAGCTCATCTGCCAGGCCACGGGTTTCAGTCCCCGGCAGATTCAGGTGTCCTGGCTGCGCGAGGGGAAGCAGGTGGGGTCTGGCGTCACCACGGACCAGGTGCAGGCTGAGGCCAAAGAGTCTGGGCCCACGACCTACAAGGTGACCAGCACACTGACCATCAAAGAGAGCGACTGGCTCGGCCAGAGCATGTTCACCTGCCGCGTGGATCACAGGGGCCTGACCTTCCAGCAGAATGCGTCCTCCATGTGTGTCCCCGATCAAGACACAGCCATCCGGGTCTTCGCCATCCCCCCATCCTTTGCCAGCATCTTCCTCACCAAGTCCACCAAGTTGACCTGCCTGGTCACAGACCTGACCACCTATGACAGCGTGACCATCTCCTGGACCCGCCAGAATGGCGAAGCTGTGAAAACCCACACCAACATCTCCGAGAGCCACCCCAATGCCACTTTCAGCGCCGTGGGTGAGGCCAGCATCTGCGAGGATGACTGGAATTCCGGGGAGAGGTTCACGTGCACCGTGACCCACACAGACCTGCCCTCGCCACTGAAGCAGACCATCTCCCGGCCCAAGGGGGTGGCCCTGCACAGGCCCGATGTCTACTTGCTGCCACCAGCCCGGGAGCAGCTGAACCTGCGGGAGTCGGCCACCATCACGTGCCTGGTGACGGGCTTCTCTCCCGCGGACGTCTTCGTGCAGTGGATGCAGAGGGGGCAGCCCTTGTCCCCGGAGAAGTATGTGACCAGCGCCCCAATGCCTGAGCCCCAGGCCCCAGGCCGGTACTTCGCCCACAGCATCCTGACCGTGTCCGAAGAGGAATGGAACACGGGGGAGACCTACACCTGCGTGGTGGCCCATGAGGCCCTGCCCAACAGGGTCACCGAGAGGACCGTGGACAAGTCCACCGAGGGGGAGGTGAGCGCCGACGAGGAGGGCTTTGAGAACCTGTGGGCCACCGCCTCCACCTTCATCGTCCTCTTCCTCCTGAGCCTCTTCTACAGTACCACCGTCACCTTGTTCAAGGTGAAA	TGTGCGAGAGCCAAACGTTGGGGGTATAGCAGCAGCTGGTGCGACTACTGG	CARAKRWGYSSSWCDYW	51	17	55	407	418	438	428	478	479	549	3982	60	T
AAACCTGGTAATAGCA-1	clonotype421	AAACCTGGTAATAGCA-1_contig_2	AGCTTCAGCTGTGGGTAGAGAAGACAGGACTCAGGACAATCTCCAGCATGGCCAGCTTCCCTCTCCTCCTCACCCTCCTCACTCACTGTGCAGGGTCCTGGGCCCAGTCTGTGCTGACTCAGCCACCCTCAGCGTCTGGGACCCCCGGGCAGAGGGTCACCATCTCTTGTTCTGGAAGCAGCTCCAACATCGGAAGTAATACTGTAAACTGGTACCAGCAGCTCCCAGGAACGGCCCCCAAACTCCTCATCTATAGTAATAATCAGCGGCCCTCAGGGGTCCCTGACCGATTCTCTGGCTCCAAGTCTGGCACCTCAGCCTCCCTGGCCATCAGTGGGCTCCAGTCTGAGGATGAGGCTGATTATTACTGTGCAGCATGGGATGACAGCCTGAATGGTGGGGTGTTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCGCCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTA	MASFPLLLTLLTHCAGSWAQSVLTQPPSASGTPGQRVTISCSGSSSNIGSNTVNWYQQLPGTAPKLLIYSNNQRPSGVPDRFSGSKSGTSASLAISGLQSEDEADYYCAAWDDSLNGGVFGGGTKLTVLGQPKAAPSVTLFPPSSEELQANKATLVCLISDFYPGAVTVAWKADSSPVKAGVETTTPSKQSNNKYAASS	T	F	IGLV1-44	47S353M246S			IGLJ3	397S38M211S	IGLC2	435S211M	AGCTTCAGCTGTGGGTAGAGAAGACAGGACTCAGGACAATCTCCAGCATGGCCAGCTTCCCTCTCCTCCTCACCCTCCTCACTCACTGTGCAGGGTCCTGGGCCCAGTCTGTGCTGACTCAGCCACCCTCAGCGTCTGGGACCCCCGGGCAGAGGGTCACCATCTCTTGTTCTGGAAGCAGCTCCAACATCGGAAGTAATACTGTAAACTGGTACCAGCAGCTCCCAGGAACGGCCCCCAAACTCCTCATCTATAGTAATAATCAGCGGCCCTCAGGGGTCCCTGACCGATTCTCTGGCTCCAAGTCTGGCACCTCAGCCTCCCTGGCCATCAGTGGGCTCCAGTCTGAGGATGAGGCTGATTATTACTGTGCAGCATGGGATGACAGCCTGAATGGTGGGGTGTTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCGCCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTA	CTGATTTGCATGGATGGACTCTCCCCCTCTCAGAGTATGAAGAGAGGGAGAGATCTGGGGGAAGCTCAGCTTCAGCTGTGGGTAGAGAAGACAGGACTCAGGACAATCTCCAGCATGGCCAGCTTCCCTCTCCTCCTCACCCTCCTCACTCACTGTGCAGGGTCCTGGGCCCAGTCTGTGCTGACTCAGCCACCCTCAGCGTCTGGGACCCCCGGGCAGAGGGTCACCATCTCTTGTTCTGGAAGCAGCTCCAACATCGGAAGTAATACTGTAAACTGGTACCAGCAGCTCCCAGGAACGGCCCCCAAACTCCTCATCTATAGTAATAATCAGCGGCCCTCAGGGGTCCCTGACCGATTCTCTGGCTCCAAGTCTGGCACCTCAGCCTCCCTGGCCATCAGTGGGCTCCAGTCTGAGGATGAGGCTGATTATTACTGTGCAGCATGGGATGACAGCCTGAATGGTCCTTGGGTGTTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCGCCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTATCTGAGCCTGACGCCTGAGCAGTGGAAGTCCCACAGAAGCTACAGCTGCCAGGTCACGCATGAAGGGAGCACCGTGGAGAAGACAGTGGCCCCTACAGAATGTTCA	TGTGCAGCATGGGATGACAGCCTGAATGGTGGGGTGTTC	CAAWDDSLNGGVF	39	13	48	400			398	435	436	646	3120	42	T
AAACCTGGTACGCACC-1	clonotype194	AAACCTGGTACGCACC-1_contig_1	ACAACCACACCCCTCCTAAGAAGAAGCCCCTAGACCACAGCTCCACACCATGGACTGGACCTGGAGGATCCTCTTCTTGGTGGCAGCAGCAACAGGTGCCCACTCCCAGGTGCAGCTGGTGCAATCTGGGTCTGAGTTGAAGAAGCCTGGGGCCTCAGTGAAGGTTTCCTGCAAGGCTTCTGGATACACCTTCACTAGCTATGCTATGAATTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAACACCAACACTGGGAACCCAACGTATGCCCAGGGCTTCACAGGACGGTTTGTCTTCTCCTTGGACACCTCTGTCAGCACGGCATATCTGCAGATCAGCAGCCTAAAGGCTGAGGACACTGCCGTGTATTACTGTGCGAGAGCCCTGGGGGCTATTGAACTCTTTGATTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	MDWTWRILFLVAAATGAHSQVQLVQSGSELKKPGASVKVSCKASGYTFTSYAMNWVRQAPGQGLEWMGWINTNTGNPTYAQGFTGRFVFSLDTSVSTAYLQISSLKAEDTAVYYCARALGAIELFDYWGQGTLVTVSSGSASAPTLFPLVSCENSPSDTSSV	T	F	IGHV7-4-1	49S353M133S			IGHJ4	416S48M71S	IGHM	464S71M	ACAACCACACCCCTCCTAAGAAGAAGCCCCTAGACCACAGCTCCACACCATGGACTGGACCTGGAGGATCCTCTTCTTGGTGGCAGCAGCAACAGGTGCCCACTCCCAGGTGCAGCTGGTGCAATCTGGGTCTGAGTTGAAGAAGCCTGGGGCCTCAGTGAAGGTTTCCTGCAAGGCTTCTGGATACACCTTCACTAGCTATGCTATGAATTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAACACCAACACTGGGAACCCAACGTATGCCCAGGGCTTCACAGGACGGTTTGTCTTCTCCTTGGACACCTCTGTCAGCACGGCATATCTGCAGATCAGCAGCCTAAAGGCTGAGGACACTGCCGTGTATTACTGTGCGAGAGCCCTGGGGGCTATTGAACTCTTTGATTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTG	ACAACCACACCCCTCCTAAGAAGAAGCCCCTAGACCACAGCTCCACACCATGGACTGGACCTGGAGGATCCTCTTCTTGGTGGCAGCAGCAACAGGTGCCCACTCCCAGGTGCAGCTGGTGCAATCTGGGTCTGAGTTGAAGAAGCCTGGGGCCTCAGTGAAGGTTTCCTGCAAGGCTTCTGGATACACCTTCACTAGCTATGCTATGAATTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGATGGATCAACACCAACACTGGGAACCCAACGTATGCCCAGGGCTTCACAGGACGGTTTGTCTTCTCCTTGGACACCTCTGTCAGCACGGCATATCTGCAGATCTGCAGCCTAAAGGCTGAGGACACTGCCGTGTATTACTGTGCGAGAGAACTACTTTGACTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGGGAGTGCATCCGCCCCAACCCTTTTCCCCCTCGTCTCCTGTGAGAATTCCCCGTCGGATACGAGCAGCGTGGCCGTTGGCTGCCTCGCACAGGACTTCCTTCCCGACTCCATCACTTTCTCCTGGAAATACAAGAACAACTCTGACATCAGCAGCACCCGGGGCTTCCCATCAGTCCTGAGAGGGGGCAAGTACGCAGCCACCTCACAGGTGCTGCTGCCTTCCAAGGACGTCATGCAGGGCACAGACGAACACGTGGTGTGCAAAGTCCAGCACCCCAACGGCAACAAAGAAAAGAACGTGCCTCTTCCAGTGATTGCTGAGCTGCCTCCCAAAGTGAGCGTCTTCGTCCCACCCCGCGACGGCTTCTTCGGCAACCCCCGCAAGTCCAAGCTCATCTGCCAGGCCACGGGTTTCAGTCCCCGGCAGATTCAGGTGTCCTGGCTGCGCGAGGGGAAGCAGGTGGGGTCTGGCGTCACCACGGACCAGGTGCAGGCTGAGGCCAAAGAGTCTGGGCCCACGACCTACAAGGTGACCAGCACACTGACCATCAAAGAGAGCGACTGGCTCGGCCAGAGCATGTTCACCTGCCGCGTGGATCACAGGGGCCTGACCTTCCAGCAGAATGCGTCCTCCATGTGTGTCCCCGATCAAGACACAGCCATCCGGGTCTTCGCCATCCCCCCATCCTTTGCCAGCATCTTCCTCACCAAGTCCACCAAGTTGACCTGCCTGGTCACAGACCTGACCACCTATGACAGCGTGACCATCTCCTGGACCCGCCAGAATGGCGAAGCTGTGAAAACCCACACCAACATCTCCGAGAGCCACCCCAATGCCACTTTCAGCGCCGTGGGTGAGGCCAGCATCTGCGAGGATGACTGGAATTCCGGGGAGAGGTTCACGTGCACCGTGACCCACACAGACCTGCCCTCGCCACTGAAGCAGACCATCTCCCGGCCCAAGGGGGTGGCCCTGCACAGGCCCGATGTCTACTTGCTGCCACCAGCCCGGGAGCAGCTGAACCTGCGGGAGTCGGCCACCATCACGTGCCTGGTGACGGGCTTCTCTCCCGCGGACGTCTTCGTGCAGTGGATGCAGAGGGGGCAGCCCTTGTCCCCGGAGAAGTATGTGACCAGCGCCCCAATGCCTGAGCCCCAGGCCCCAGGCCGGTACTTCGCCCACAGCATCCTGACCGTGTCCGAAGAGGAATGGAACACGGGGGAGACCTACACCTGCGTGGTGGCCCATGAGGCCCTGCCCAACAGGGTCACCGAGAGGACCGTGGACAAGTCCACCGAGGGGGAGGTGAGCGCCGACGAGGAGGGCTTTGAGAACCTGTGGGCCACCGCCTCCACCTTCATCGTCCTCTTCCTCCTGAGCCTCTTCTACAGTACCACCGTCACCTTGTTCAAGGTGAAA	TGTGCGAGAGCCCTGGGGGCTATTGAACTCTTTGATTACTGG	CARALGAIELFDYW	42	14	50	402			417	464	465	535	3554	61	T
AAACCTGGTACGCACC-1	clonotype194	AAACCTGGTACGCACC-1_contig_2	TCTGGCACCAGGGGTCCCTTCCAATATCAGCACCATGGCCTGGACTCCTCTCTTTCTGTTCCTCCTCACTTGCTGCCCAGGGTCCAATTCCCAGGCTGTGGTGACTCAGGAGCCCTCACTGACTGTGTCCCCAGGAGGGACAGTCACTCTCACCTGTGGCTCCAGCACTGGAGCTGTCACCAGTGGTCATTATCCCTACTGGTTCCAGCAGAAGCCTGGCCAAGCCCCCAGGACACTGATTTATGATACAAGCAACAAACACTCCTGGACCCCTGCCCGGTTCTCAGGCTCCCTCCTTGGGGGCAAAGCTGCCCTGACCCTTTCGGGTGCGCAGCCTGAGGATGAGGCTGAGTATTACTGCTTGCTCTCCTATAGTGGTGCCCATGTGGTATTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCGCCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTA	MAWTPLFLFLLTCCPGSNSQAVVTQEPSLTVSPGGTVTLTCGSSTGAVTSGHYPYWFQQKPGQAPRTLIYDTSNKHSWTPARFSGSLLGGKAALTLSGAQPEDEAEYYCLLSYSGAHVVFGGGTKLTVLGQPKAAPSVTLFPPSSEELQANKATLVCLISDFYPGAVTVAWKADSSPVKAGVETTTPSKQSNNKYAASS	T	F	IGLV7-46	34S351M248S			IGLJ2	384S38M211S	IGLC2	422S211M	TCTGGCACCAGGGGTCCCTTCCAATATCAGCACCATGGCCTGGACTCCTCTCTTTCTGTTCCTCCTCACTTGCTGCCCAGGGTCCAATTCCCAGGCTGTGGTGACTCAGGAGCCCTCACTGACTGTGTCCCCAGGAGGGACAGTCACTCTCACCTGTGGCTCCAGCACTGGAGCTGTCACCAGTGGTCATTATCCCTACTGGTTCCAGCAGAAGCCTGGCCAAGCCCCCAGGACACTGATTTATGATACAAGCAACAAACACTCCTGGACCCCTGCCCGGTTCTCAGGCTCCCTCCTTGGGGGCAAAGCTGCCCTGACCCTTTCGGGTGCGCAGCCTGAGGATGAGGCTGAGTATTACTGCTTGCTCTCCTATAGTGGTGCCCATGTGGTATTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCGCCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTA	TCTGGCACCAGGGGTCCCTTCCAATATCAGCACCATGGCCTGGACTCCTCTCTTTCTGTTCCTCCTCACTTGCTGCCCAGGGTCCAATTCCCAGGCTGTGGTGACTCAGGAGCCCTCACTGACTGTGTCCCCAGGAGGGACAGTCACTCTCACCTGTGGCTCCAGCACTGGAGCTGTCACCAGTGGTCATTATCCCTACTGGTTCCAGCAGAAGCCTGGCCAAGCCCCCAGGACACTGATTTATGATACAAGCAACAAACACTCCTGGACACCTGCCCGGTTCTCAGGCTCCCTCCTTGGGGGCAAAGCTGCCCTGACCCTTTTGGGTGCGCAGCCTGAGGATGAGGCTGAGTATTACTGCTTGCTCTCCTATAGTGGTGCTCGGTGTGGTATTCGGCGGAGGGACCAAGCTGACCGTCCTAGGTCAGCCCAAGGCTGCCCCCTCGGTCACTCTGTTCCCGCCCTCCTCTGAGGAGCTTCAAGCCAACAAGGCCACACTGGTGTGTCTCATAAGTGACTTCTACCCGGGAGCCGTGACAGTGGCCTGGAAGGCAGATAGCAGCCCCGTCAAGGCGGGAGTGGAGACCACCACACCCTCCAAACAAAGCAACAACAAGTACGCGGCCAGCAGCTATCTGAGCCTGACGCCTGAGCAGTGGAAGTCCCACAGAAGCTACAGCTGCCAGGTCACGCATGAAGGGAGCACCGTGGAGAAGACAGTGGCCCCTACAGAATGTTCA	TGCTTGCTCTCCTATAGTGGTGCCCATGTGGTATTC	CLLSYSGAHVVF	36	12	35	385			385	422	423	633	3024	38	T
AAACCTGTCCAACCAA-1	clonotype136	AAACCTGTCCAACCAA-1_contig_1	GAGCTACAACAGGCAGGCAGGGGCAGCAAGATGGTGTTGCAGACCCAGGTCTTCATTTCTCTGTTGCTCTGGATCTCTGGTGCCTACGGGGACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGGGCGAGAGGGCCACCATCAACTGCAAGTCCAGCCAGAGTGTTTTATACAGCTCCAACAATAAGAACTACTTAGCTTGGTACCAGCAGAAACCAGGACAGCCTCCTAAGCTGCTCATTTACTGGGCATCTACCCGGGAATCCGGGGTCCCTGACCGATTCAGTGGCAGCGGGTCTGGGACAGATTTCACTCTCACCATCAGCAGCCTGCAGGCTGAAGATGTGGCAGTTTATTACTGTCAGCAATATTATAGTACTCGGCGGACGTTCGGCCAAGGGACCAAGGTGGAAATCAAACGAACTGTGGCTGCACCATCTGTCTTCATCTTCCCGCCATCTGATGAGCAGTTGAAATCTGGAACTGCCTCTGTTGTGTGCCTGCTGAATAACTTCTATCCCAGAGAGGCCAAAGTACAGTGGAAGGTGGATAACGC	MVLQTQVFISLLLWISGAYGDIVMTQSPDSLAVSLGERATINCKSSQSVLYSSNNKNYLAWYQQKPGQPPKLLIYWASTRESGVPDRFSGSGSGTDFTLTISSLQAEDVAVYYCQQYYSTRRTFGQGTKVEIKRTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDN	T	F	IGKV4-1	30S363M173S			IGKJ1	392S38M136S	IGKC	430S136M	GAGCTACAACAGGCAGGCAGGGGCAGCAAGATGGTGTTGCAGACCCAGGTCTTCATTTCTCTGTTGCTCTGGATCTCTGGTGCCTACGGGGACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGGGCGAGAGGGCCACCATCAACTGCAAGTCCAGCCAGAGTGTTTTATACAGCTCCAACAATAAGAACTACTTAGCTTGGTACCAGCAGAAACCAGGACAGCCTCCTAAGCTGCTCATTTACTGGGCATCTACCCGGGAATCCGGGGTCCCTGACCGATTCAGTGGCAGCGGGTCTGGGACAGATTTCACTCTCACCATCAGCAGCCTGCAGGCTGAAGATGTGGCAGTTTATTACTGTCAGCAATATTATAGTACTCGGCGGACGTTCGGCCAAGGGACCAAGGTGGAAATCAAACGAACTGTGGCTGCACCATCTGTCTTCATCTTCCCGCCATCTGATGAGCAGTTGAAATCTGGAACTGCCTCTGTTGTGTGCCTGCTGAATAACTTCTATCCCAGAGAGGCCAAAGTACAGTGGAAGGTGGATAACGC	TTTGGCTCTTGATTTACATTGGGTACTTTCACAACCCACTGCTCATGAAATTTGCTTTTGTACTCACTGGTTGTTTTTGCATAGGCCCCTCCAGGCCACGACCAGCTGTTTGGATTTTATAAACGGGCCGTTTGCATTGTGAACTGAGCTACAACAGGCAGGCAGGGGCAGCAAGATGGTGTTGCAGACCCAGGTCTTCATTTCTCTGTTGCTCTGGATCTCTGGTGCCTACGGGGACATCGTGATGACCCAGTCTCCAGACTCCCTGGCTGTGTCTCTGGGCGAGAGGGCCACCATCAACTGCAAGTCCAGCCAGAGTGTTTTATACAGCTCCAACAATAAGAACTACTTAGCTTGGTACCAGCAGAAACCAGGACAGCCTCCTAAGCTGCTCATTTACTGGGCATCTACCCGGGAATCCGGGGTCCCTGACCGATTCAGTGGCAGCGGGTCTGGGACAGATTTCACTCTCACCATCAGCAGCCTGCAGGCTGAAGATGTGGCAGTTTATTACTGTCAGCAATATTATAGTACTCCTGTGGACGTTCGGCCAAGGGACCAAGGTGGAAATCAAACGAACTGTGGCTGCACCATCTGTCTTCATCTTCCCGCCATCTGATGAGCAGTTGAAATCTGGAACTGCCTCTGTTGTGTGCCTGCTGAATAACTTCTATCCCAGAGAGGCCAAAGTACAGTGGAAGGTGGATAACGCCCTCCAATCGGGTAACTCCCAGGAGAGTGTCACAGAGCAGGACAGCAAGGACAGCACCTACAGCCTCAGCAGCACCCTGACGCTGAGCAAAGCAGACTACGAGAAACACAAAGTCTACGCCTGCGAAGTCACCCATCAGGGCCTGAGCTCGCCCGTCACAAAGAGCTTCAACAGGGGAGAGTGT	TGTCAGCAATATTATAGTACTCGGCGGACGTTC	CQQYYSTRRTF	33	11	31	393			393	430	431	566	4259	46	T
```

# Adaptive Immune Receptor Repertoire analysis

[nf-core/airrflow](https://nf-co.re/airrflow/4.2.0/):

> nf-core/airrflow is a bioinformatics best-practice pipeline to analyze B-cell or T-cell repertoire sequencing data. It makes use of the Immcantation toolset. The input data can be targeted amplicon bulk sequencing data of the V, D, J and C regions of the B/T-cell receptor with multiplex PCR or 5’ RACE protocol, single-cell VDJ sequencing using the 10xGenomics libraries, or assembled reads (bulk or single-cell).
>
> nf-core/airrflow allows the end-to-end processing of BCR and TCR bulk and single cell targeted sequencing data. Several protocols are supported, please see the usage documentation for more details on the supported protocols. The pipeline has been certified as AIRR compliant by the AIRR community, which means that it is compatible with downstream analysis tools also supporting this format.

## Setup

1. Install `nf-core/tools`

```console
pip install --upgrade nf-core
nf-core --version
```
```
                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 3.2.0 - https://nf-co.re


nf-core, version 3.2.0
```

2. Find and download `nf-core/airrflow`

```console
nf-core pipelines list | grep airrflow
```
```
                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 3.2.0 - https://nf-co.re


│ airrflow                  │    59 │          4.2.0 │  3 months ago │           - │ -                    │
```

Download.

```console
export NXF_SINGULARITY_CACHEDIR=/data/nxf_singularity
nf-core pipelines download --revision 4.2.0 --outdir /data/nf-core/airrflow --compress none --container-system singularity --parallel-downloads 4 airrflow
# select amend
```

3. Test

```console
nextflow run /data/nf-core/airrflow/4_2_0/main.nf -profile test,singularity --outdir /data/airrflow_test
```
```
-[nf-core/airrflow] Pipeline completed successfully-
Completed at: 08-Mar-2025 16:47:51
Duration    : 10m 40s
CPU hours   : 0.7
Succeeded   : 221
```

## Test results

Results were written to `/data/airrflow_test`. See the [nf-core/airrflow output page](https://nf-co.re/airrflow/4.2.0/docs/output/) for more information.

```console
tree -L 1 --charset ascii /data/airrflow_test
```
```
/data/airrflow_test
|-- Airrflow_report.html
|-- clonal_analysis
|-- fastp
|-- fastqc
|-- multiqc
|-- parsed_logs
|-- pipeline_info
|-- presto
|-- qc-filtering
|-- repertoire_comparison
|-- report_file_size
`-- vdj_annotation
```

## Input samplesheet

TSV format:

* `sample_id`: Sample ID assigned by submitter, unique within study.
* `filename_R1`: path to fastq file with first mates of paired-end sequencing.
* `filename_R2`: path to fastq file with second mates of paired-end sequencing.
* `filename_I`1 (optional): path to fastq with illumina index and UMI (unique molecular identifier) barcode.
* `subject_id`: Subject ID assigned by submitter, unique within study.
* `species`: species from which the sample was taken. Supported species are human and mouse.
* `tissue`: tissue from which the sample was taken. E.g. blood, PBMC, brain.
* `pcr_target_locus`: Designation of the target locus (`IG` or `TR`).
* `biomaterial_provider`: Institution / research group that provided the samples.
* `sex`: Subject biological sex (female, male, etc.).
* `age`: Subject biological age.
* `single_cell`: TRUE or FALSE.

### Bulk

The [required input file](https://nf-co.re/airrflow/usage#fastq-input-samplesheet-bulk-airr-sequencing) for processing raw BCR or TCR bulk targeted sequencing data is a sample sheet in TSV format (tab separated). The following columns are required:

* `sample_id`
* `filename_R1`
* `filename_R2`
* `subject_id`
* `species`
* `tissue`
* `pcr_target_locus`
* `single_cell`
* `sex`
* `age`
* `biomaterial_provider`

| sample_id | filename_R1                     | filename_R2                     | filename_I1                     | subject_id | species | pcr_target_locus | tissue | sex    | age | biomaterial_provider | single_cell | intervention   | collection_time_point_relative | cell_subset  |
| -         | -                               | -                               | -                               | -          | -       | -                | -      | -      | -   | -                    | -           | -              | -                              | -            |
| sample01  | sample1_S8_L001_R1_001.fastq.gz | sample1_S8_L001_R2_001.fastq.gz | sample1_S8_L001_I1_001.fastq.gz | Subject02  | human   | IG               | blood  | NA     | 53  | sequencing_facility  | FALSE       | Drug_treatment | Baseline                       | plasmablasts |
| sample02  | sample2_S8_L001_R1_001.fastq.gz | sample2_S8_L001_R2_001.fastq.gz | sample2_S8_L001_I1_001.fastq.gz | Subject02  | human   | TR               | blood  | female | 78  | sequencing_facility  | FALSE       | Drug_treatment | Baseline                       | plasmablasts |

### Single cell

The [required input file](https://nf-co.re/airrflow/usage#fastq-input-samplesheet-single-cell-sequencing) for processing raw BCR or TCR single cell targeted sequencing data is a sample sheet in TSV format (tab separated). The columns are required:

* `sample_id`
* `filename_R1`
* `filename_R2`
* `subject_id`
* `species`
* `tissue`
* `pcr_target_locus`
* `single_cell`
* `sex`
* `age`
* `biomaterial_provider`

Any other columns you add will be available in the final repertoire file as extra metadata fields. You can refer to the bulk fastq input section for documentation on the individual columns. An example samplesheet is:

| sample_id | filename_R1                      | filename_R2                      | subject_id | species | pcr_target_locus | tissue | sex    | age | biomaterial_provider | single_cell |
| -         | -                                | -                                | -          | -       | -                | -      | -      | -   | -                    | -           |
| sample01  | sample01_S1_L001_R1_001.fastq.gz | sample01_S1_L001_R2_001.fastq.gz | Subject02  | human   | IG               | blood  | NA     | 53  | sequencing_facility  | TRUE        |
| sample02  | sample02_S1_L001_R1_001.fastq.gz | sample02_S1_L001_R2_001.fastq.gz | Subject02  | human   | TR               | blood  | female | 78  | sequencing_facility  | TRUE        |


FASTQ files must conform with the 10x Genomics Cell Ranger naming conventions with the same sample name as provided in the `sample_id` column:

```
>**[SAMPLE-NAME] S[CHIP-NUMBER]_ L00[LANE-NUMBER]_[R1/R2]_001.fastq.gz**
```

Read type is one of

* I1: Sample index read (optional)
* I2: Sample index read (optional)
* R1: Read 1
* R2: Read 2

It is possible to provide several FASTQ files per sample (e.g. sequenced over different chips or lanes). In this case the different FASTQ files per sample will be provided to the same Cell Ranger process. These rows should then have an identical `sample_id` field.

### Assembled input samplesheet

* The required input file for processing raw BCR or TCR bulk targeted sequencing data is a sample sheet in TSV format (tab separated). The following columns are required:
    * `sample_id`
    * `filename`
    * `subject_id`
    * `species`
    * `tissue`
    * `single_cell`
    * `sex`
    * `age`
    * `biomaterial_provider`

All fields are explained in the previous section, with the only difference being that there is only one filename column for the assembled input samplesheet. The provided file will be different from assembled single-cell or bulk data:

* `filename` for single-cell assembled data: path to `airr_rearrangement.tsv` file, for example the one generated when processing the 10x Genomics scBCRseq / scTCRseq with 10x Genomics cellranger `cellranger vdj` or `cellranger multi`.
    * The field accepts any tsv tables following the AIRR rearrangement Schema specification.
* `filename` for bulk assembled data: path to `sequences.fasta` file, containing the assembled and error-corrected reads.

The required input file for processing raw BCR or TCR bulk targeted sequencing data is a sample sheet in TSV format (tab separated). The columns `sample_id`, `filename`, `subject_id`, `species`, `tissue`, `single_cell`, `pcr_target_locus`, `sex`, `age` and `biomaterial_provider` are required.

An example samplesheet is:

| filename                                    | species | subject_id | sample_id              | tissue | sex  | age | biomaterial_provider | pcr_target_locus | single_cell |
| -                                           | -       | -          | -                      | -      | -    | -   | -                    | -                | -           |
| sc5p_v2_hs_PBMC_1k_b_airr_rearrangement.tsv | human   | subject_x  | sc5p_v2_hs_PBMC_1k_5fb | PBMC   | NA   | NA  | 10x Genomics         | IG               | TRUE        |
| bulk-Laserson-2014.fasta                    | human   | PGP1       | PGP1                   | PBMC   | male | NA  | Laserson-2014        | IG               | FALSE       |

### Troubleshooting

The process `NFCORE_AIRRFLOW:AIRRFLOW:CLONAL_ANALYSIS:FIND_CLONAL_THRESHOLD` requires 16 CPUs by default.

```console
cat error.log
```
```
Mar-22 22:35:13.429 [Task submitter] ERROR nextflow.processor.TaskProcessor - Error executing process > 'NFCORE_AIRRFLOW:AIRRFLOW:CLONAL_ANALYSIS:FIND_CLONAL_THRESHOLD (all_reps)'

Caused by:
  Process requirement exceeds available CPUs -- req: 16; avail: 6


Command executed:

  Rscript -e "enchantr::enchantr_report('find_threshold', \
          report_params=list('input'='find_threshold_samplesheet.txt',\
              'cloneby'='subject_id',\
              'crossby'='subject_id',\
              'singlecell'='single_cell',\
              'outdir'=getwd(),\
              'nproc'=16,\
              'outname'='all_reps',\
              'log'='all_reps_threshold_command_log',\
              'logo'='nf-core-airrflow_logo_reports.png' ,'findthreshold_cutoff'='user','findthreshold_edge'=0.9,'findthreshold_method'='gmm','findthreshold_model'='gamma-norm','findthreshold_spc'=0.995))"

      cp -r enchantr all_reps_dist_report && rm -rf enchantr

      echo "NFCORE_AIRRFLOW:AIRRFLOW:CLONAL_ANALYSIS:FIND_CLONAL_THRESHOLD": > versions.yml
      Rscript -e "cat(paste0('  enchantr: ',packageVersion('enchantr'),'
  '))" >> versions.yml

Command exit status:
  -

Command output:
  (empty)
```

After some searching, the script is run via the Nextflow script `find_threshold.nf`.

```nextflow
    script:
    def args = task.ext.args ? asString(task.ext.args) : ''
    """
    Rscript -e "enchantr::enchantr_report('find_threshold', \\
        report_params=list('input'='${tabs_samplesheet}',\\
            'cloneby'='${params.cloneby}',\\
            'crossby'='${params.crossby}',\\
            'singlecell'='${params.singlecell}',\\
            'outdir'=getwd(),\\
            'nproc'=${task.cpus},\\
            'outname'='all_reps',\\
            'log'='all_reps_threshold_command_log',\\
            'logo'='${logo}' ${args}))"

    cp -r enchantr all_reps_dist_report && rm -rf enchantr

    echo "${task.process}": > versions.yml
    Rscript -e "cat(paste0('  enchantr: ',packageVersion('enchantr'),'\n'))" >> versions.yml
    """
```

* Set `-process.cpus=4` but got the same error.
* Searched again and found that some CPU settings are set in `base.config` and noticed:

```nextflow
withLabel:process_long_parallelized {
    time   = { 30.h  * task.attempt }
    cpus   = { 16    * task.attempt }
    memory = { 72.GB * task.attempt }
}
```

* Looked again at `find_threshold.nf` and saw the label:

```nextflow
process FIND_THRESHOLD {
    tag "all_reps"

    label 'process_long_parallelized'
```

* Manually changed `base.config` to 4 cpus.
* That worked but got the new error: `  Process requirement exceeds available memory -- req: 72 GB; avail: 62.6 GB`
* Manually changed `base.config` to 52.GB.

New error:

```
ERROR ~ Error executing process > 'NFCORE_AIRRFLOW:AIRRFLOW:REPERTOIRE_ANALYSIS_REPORTING:AIRRFLOW_REPORT (all_reps)'

Caused by:
  Process requirement exceeds available CPUs -- req: 12; avail: 6
```

* Looked in `airrflow_report.nf` and saw that it uses `label 'process_high'`
* Manually changed `process_high` to 4 cpus and 52.GB

Finally `-[nf-core/airrflow] Pipeline completed successfully-`.

(Final note: removed `-process.cpus=4` from `./nfcore_airrflow.sh` and the workflow could still successfully complete.)
